crates/astria-sequencer/src/slinky/oracle/state_ext.rs

```diff
@@ -0,0 +1,278 @@
+use std::collections::HashMap;
+
+use anyhow::{
+    bail,
+    Context,
+    Result,
+};
+use astria_core::slinky::{
+    oracle::v1::{
+        CurrencyPairState,
+        QuotePrice,
+    },
+    types::v1::CurrencyPair,
+};
+use async_trait::async_trait;
+use borsh::{
+    BorshDeserialize,
+    BorshSerialize,
+};
+use cnidarium::{
+    StateRead,
+    StateWrite,
+};
+use futures::StreamExt as _;
+use tracing::instrument;
+
+const CURRENCY_PAIR_TO_ID_PREFIX: &str = "oraclecpid";
+const ID_TO_CURRENCY_PAIR_PREFIX: &str = "oracleidcp";
+const CURRENCY_PAIR_STATE_PREFIX: &str = "oraclecpstate";
+
+const NUM_CURRENCY_PAIRS_KEY: &str = "oraclenumcps";
+const NUM_REMOVED_CURRENCY_PAIRS_KEY: &str = "oraclenumremovedcps";
+const NEXT_CURRENCY_PAIR_ID_KEY: &str = "oraclenextcpid";
+
+fn currency_pair_to_id_storage_key(currency_pair: &CurrencyPair) -> String {
+    format!("{CURRENCY_PAIR_TO_ID_PREFIX}/{currency_pair}",)
+}
+
+fn id_to_currency_pair_storage_key(id: u64) -> String {
+    format!("{ID_TO_CURRENCY_PAIR_PREFIX}/{id}")
+}
+
+fn currency_pair_state_storage_key(currency_pair: &CurrencyPair) -> String {
+    format!("{CURRENCY_PAIR_STATE_PREFIX}/{currency_pair}",)
+}
+
+/// Newtype wrapper to read and write a u64 from rocksdb.
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+struct Id(u64);
+
+/// Newtype wrapper to read and write a u64 from rocksdb.
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+struct Count(u64);
+
+#[async_trait]
+pub(crate) trait StateReadExt: StateRead {
+    #[instrument(skip_all)]
+    async fn get_currency_pair_id(&self, currency_pair: &CurrencyPair) -> Result<u64> {
+        let Some(bytes) = self
+            .get_raw(&currency_pair_to_id_storage_key(currency_pair))
+            .await
+            .context("failed reading currency pair id from state")?
+        else {
+            bail!("currency pair not found in state")
+        };
+        let Id(id) = Id::try_from_slice(&bytes).context("invalid currency pair id bytes")?;
+        Ok(id)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_currency_pair(&self, id: u64) -> Result<Option<CurrencyPair>> {
+        let bytes = self
+            .get_raw(&id_to_currency_pair_storage_key(id))
+            .await
+            .context("failed to get currency pair from state")?;
+        match bytes {
+            Some(bytes) => {
+                let currency_pair = serde_json::from_slice(&bytes)
+                    .context("failed to deserialize currency pair")?;
+                Ok(Some(currency_pair))
+            }
+            None => Ok(None),
+        }
+    }
+
+    #[instrument(skip_all)]
+    async fn get_currency_pair_mapping(&self) -> Result<HashMap<u64, CurrencyPair>> {
```

Comment:
Instead of returning an object, implement this as a stream returning Items of type to make this more flexible `Result<CurrencyPair>` (or whatever). See https://github.com/astriaorg/astria/pull/1453 for how this can be done (this was ultimately merged into https://github.com/astriaorg/astria/pull/1408)